manageiq_config.py - add ability to modify ManageIQ config

itewk opened the pull request:
##### SUMMARY
Adds the ability to modify ManageIQ config. The ability to manage the config of a ManageIQ appliance is not exposed via the REST API therefor this module uses the rails runner to run the `MiqServer.my_server.get_config` and `MiqServer.my_server.set_config` functions to acomplish it's task.

This module is very important to being able to fully automate the deployment and configuration management of ManageIQ appliances because otherwise a user has to manually log into the UI of each and every appliance and go to the advanced settings tab to modify settings like memory thresholds and SMTP configuration.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
manageiq_config

##### ANSIBLE VERSION
```
ansible 2.4.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/ian/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, May  3 2017, 07:55:04) [GCC 4.8.5 20150623 (Red Hat 4.8.5-14)]

```

##### ADDITIONAL INFORMATION
```
# set the smtp settings
- manageiq_config:
    name: smtp
    value:
      from: cfme@example.com
      host: postfix.example.com
      port: 25
      domain: example.com

# set generic worker memory threshold
- manageiq_config:
    name: workers
    value:
      worker_base:
        queue_worker_base:
          generic_worker:
            memory_threshold: '900.megabytes'
```

itewk committed:
manageiq_config.py - init commit. Module for managing ManageIQ configuration.

yaacov commented:
cc @elad661

elad661 commented:
Let me know when you finish addressing the bot's feedback so I can start doing a proper review.

I think this kind of module is important, but the implementation has a major downside - it doesn't seem possible to use this on "podified ManageIQ" (ie. ManageIQ running in a container under OpenShift).

It would've been much better if API could be added for this functionality.

yaacov commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,226 @@
+    # verifty that the updated value equals the expected value
+    #     if the updated value does not match the expected value then error
+    #     else update or noop succesful
+    if update_value != updated_value:
```
have you ever reached this `update_value != updated_value` ?

the while in L192 will run until `retries >= CONFIRM_UPDATE_MAX_RETRY` is true or `fail`
so if `update_value != updated_value` it must be also `retries >= CONFIRM_UPDATE_MAX_RETRY` and
then you will `fail` in L202 ?

am I missing something ?

itewk replied:
@yaacov you are not missing something. I missed it. Originally i didn't have the retry loop in there and then after I added it in I didn't notice the logic redundancy you pointed out. Thank you. Latest commit removes the redundant if/else

yaacov commented:
shipit

itewk commented:
@elad661 Bot is now happy.

I agree, if there was MIQ API exposed for these settings that would be great. Maybe next i'll go to MIQ and see if I can add things to the API. Though I think all of the API to this point is distributed and these specific settings are appliance specific, so could be tricky.

elad661 commented:
@itewk Getting this kind of API into ManageIQ is definitely not going to be an easy task. I suggest you also contact the API team to ask if they have plans for such API for one of the next versions. I think I heard that it might be planned, but I don't know if that information is accurate.

I guess that, in theory, we could accept this module now and modify it to use the API if such API ever gets committed into ManageIQ.

I'll try to do a full review tomorrow morning.

itewk commented:
@elad661 My thoughts exactly. I would certainly be happy to update to use API when/if it ever becomes available. Your review at this time would be most welcome.

I warn you this is my first Python development, been doing mostly ruby lately and before that was all JavaScript and before that Java for Eclipse WTP project. So all those languages floating around up there can cause my style to be a bit erratic compared to language purists. Feedback is a gift and is mostly welcome and appreciated.

itewk commented:
I did discover today that in a hardened environment the `sudo_flags` value in `ansible.cfg` needs to include `-i` otherwise this will not work. Not sure if that is something the module should handle or document somehow so the next person doesn't have to spend 3 hours fighting with `ruby: error while loading shared libraries: libruby.so.2.3: cannot open shared object file: No such file or directory` errors from the `rails r` calls.

yaacov commented:
> Not sure if that is something the module should handle or document somehow

a. should be documented in the description ( unless you do not like the users )  :-)
https://github.com/ansible/ansible/pull/35182/files#diff-63f6bb4f3305ae65e40aa6e13780a0a5R22

b. _if you can_ handle it in the module, and document inline why you did it ...

itewk commented:
@yaacov any ideas on how to handle i in the module? I don't know enough about the magic of how the appliance normally sets up the environment when you log onto the appliance to emulate it when SUDOing to root on an appliance where the environment isn't carried through by default or by updating `sudo_flags`. You know anything about what CFME does to set up the environment when a user logs in so I can manually run that to ensure the environment is set up correctly?

itewk commented:
@yaacov I have been fighting with trying to do this automatically programaticlly.

I was able to get the following to work for reading in the existing parameters:
```
(get_value_rc, get_value_out, get_value_err) = module.run_command(
    ". /etc/profile && rails r 'puts MiqServer.my_server.get_config(:%s).config.to_json'" % (name),
    use_unsafe_shell=True,
    cwd=VMDB_PATH
)
```

But I can not for the life of me get the writing to work due to to many levels of nested quotes. The issue is needing to wrap the argument to `rails r` in quotes, and then wrap the argument to `JSON.parse` in quotes, and then the result of `json.dumps` being all double quoated. When using the run_command as a command not shell it worked passing each section of the command as part of the array, but having to pass the command all as one large command to shell, my brain is struggling with all the quote escaping.

```
(update_value_rc, update_value_out, update_value_err) = module.run_command(
    ". /etc/profile && rails r 'MiqServer.my_server.set_config(:%s => JSON.parse('%s'))'" % (VMDB_PATH, module.params['name'], json.dumps(update_value)),
    cwd=VMDB_PATH,
    use_unsafe_shell=True
)
```

That is not to mention that this approach requires running in `use_unsafe_shell` mode and is noticeably slower. I was trying to figure out some way of looking at the `env` to decide if the environment is correct or not and then programmatically deciding whether to run `. /etc/profile` and using the shell.  But then that went down a rabbit whole of if i should use `module.run_command('env')` or if there was some way to look at `ansible_env` but then realizing `ansible_env` isn't set if `gather_facts` was false. I was also playing with the idea of like figuring out which part of the environment is required and hard coding that into the module to set the environment as argument to `run_command` but then if miq changes the standard environment the module would need to keep up with that. So then having to handle that case. I also thought about just adding a parameter that was like `reload_profile` that a caller could set to `True` if running in an environment where `sudo -i` is disabled. But then exposing that level of complexity seemed worse on the user. But then again requiring `sudo -i` isn't awesome either. And of course all of that is mute if I can't figure out the nested quoting.

Basically, I am going around in circles of complexity. Some out side opinions at this point are most welcome.

yaacov commented:
> Some out side opinions at this point are most welcome.

my opinion is don't, documenting that user need to set up X and Y before using the module is enough for me.

@cben ?

p.s.
@cben I think we have a script that set up env vars for podified, do you remember something ?

elad661 commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,215 @@
+    try:
+        original_value = get_manageiq_config_value(module, module.params['name'])
+    except Exception as err:
+        module.fail_json(msg=str(err), **result)
```
Why do you use `result` here? It won't contain any useful information. Also, it's better to give the user some more informative error message. See https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/remote_management/manageiq/manageiq_alert_profiles.py#L102 for an example

itewk replied:
the err message has a useful description as defined in the Exception that is thrown.

elad661 commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,215 @@
+    return json.loads(get_value_out)
+
+
+def run_module():
```
This function is quite big, and it does a lot of things. Might be better to split to to smaller functions, like we do in our other modules.

itewk replied:
attempted to break it up somewhat. if you have any ideas for anything else to break out let me know. My default is only to break into a function if it is something that will be called more then once.

elad661 commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,215 @@
+        except Exception as err:
+            module.fail_json(msg=str(err), **result)
+    else:
+        updated_value = original_value
```
You probably should just exit here with `changed=False`

cben commented:
@abellotti @Fryguy We need a settings API, this is pretty round-about :-(
This module does deep-merge in order to call (via ansible ssh and `rails runner`)
https://github.com/ManageIQ/manageiq/blob/3921e8791/app/models/miq_server/configuration_management.rb#L16
which calls
https://github.com/ManageIQ/manageiq/blob/3921e8791/lib/vmdb/settings.rb#L46
which does deep-diff in order to figure out which `settings_changes` DB rows to write.

IMHO we need a settings API that supports patch semantics: you send only changes you want, not whole thing.  [Bonus point if it includes a way to delete a customization, going back to inheriting it from yml files.]

We have https://github.com/ManageIQ/manageiq-api/blob/master/app/controllers/api/settings_controller.rb ,
https://github.com/ManageIQ/manageiq-api/blob/master/spec/requests/settings_spec.rb
IIUC that's read-only.

elad661 commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,215 @@
+            module.fail_json(msg=str(err), **result)
+
+    # error if could not confirm update in timeout
+    if update_value != updated_value and retries >= CONFIRM_UPDATE_MAX_RETRY:
```
the `and retries >= CONFIRM_UPDATE_MAX_RETRY` is redundant here. If the execution got here, it'll always be `True`

itewk replied:
okay.

elad661 commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,215 @@
+    else:
+        updated_value = original_value
+
+    # setting the configuration is an async operation which must be waited for for completion
```
Is there a way to get a callback from inside the rails runner, and make the rails runner block until the callback is called?

itewk replied:
not that i have found.

elad661 commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,215 @@
+            "rails",
+            "r",
+            "MiqServer.my_server.set_config(:%s => JSON.parse('%s'))" % (module.params['name'], json.dumps(update_value))
+        ], cwd=VMDB_PATH)
```
If we do go with this approach, I feel like `VMDB_PATH` should be user-configurable

elad661 commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,215 @@
+    :param name: ManageIQ config option name
+    :return: dict of value of the ManageIQ config option
+    """
+    (get_value_rc, get_value_out, get_value_err) = module.run_command([
```
It'll look more clean if you name these `returncode`, `out`, and `_` (the underscore denotes you're not going to use that, and it's a common pattern in Python). You can also drop the parenthesis, they're not needed

itewk replied:
okay, though using `get_value_err` in the error case so will change that to `err` instead of `-`

elad661 commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,215 @@
+
+
+def main():
+    run_module()
```
having `main()` call `run_module()` feels a bit redundant, I see no reason to not put all of the main logic in `main` (and splitting actions into functions where it makes sense)

itewk replied:
I got that from http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html 

```
def main():
    run_module()

if __name__ == '__main__':
    main()
```

I will go ahead and change it but I suggest the docs get updated if that is not a valid example.

elad661 commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,215 @@
+import copy
+
+VMDB_PATH = '/var/www/miq/vmdb'
+CONFIRM_UPDATE_MAX_RETRY = 10
```
This should be configurable in the module arguments.

itewk replied:
okay.

cben commented:
I don't have an opinion whether this command-line based implementation is worth having, deferring to @itewk @yaacov.  Some tips though:
- if possible, bypass shell quoting (don't use_unsafe_shell)
- if you must, see `shlex.quote` aka `pipes.quote` https://stackoverflow.com/a/6874028/239657
- maybe you can bypass ruby quoting by passing json as another arg or stdin (and having ruby look at that)
- perhaps using https://github.com/ManageIQ/manageiq/blob/master/tools/configure_server_settings.rb or https://github.com/ManageIQ/manageiq/blob/master/lib/tasks/evm_settings.rake would help?

elad661 reviewed (commented):
I've only done a quick review of the code and haven't actually tested it yet, but I must say I'm not sure if it's a good idea to use this approach of using the rails runner, it feels a bit too fragile.

This module is very much needed, and I appreciate the effort you've made in writing it, but when something is in upstream ansible/ansible people expect it to work reliably and consistently across different ManageIQ versions - and we can't really guarantee that when using the rails runner approach.

Another thing worth thinking about is that if we merge it now, and an API will be added later, we'll have to change the module args, and that would break people's playbooks.

yaacov commented:
@itewk @cben @abellotti 

a. we need a way to modify settings.
b. no api at the moment.
c. this module is a hack ...

maybe it will be better to put it in [ansible galaxy](https://galaxy.ansible.com/) for now, and write a new module when we have a stable api ?

itewk commented:
@elad661 I think I addressed all your helpful feedback. Thanks again for taking the time to do the review.

henrygrover commented:
I agree it's not ideal that this can't work with CloudForms on OCP, but would argue that since there are no better options for that situation, and since the majority of CloudForms installs are still appliance VMs, it's better to upstream this and switch it to use an API later if one comes to exist. It addresses a major need, and I haven't heard anything about a forthcoming solution from any other source.

Tested it out in my environment and it seems to work great, thanks @itewk. Caught some broken NTP settings that I had previously missed on one appliance.

yaacov commented:
@cben @elad661 , please re-visit this issue.

IMHO , if thais scratch an itch @henrygrover and @itewk ( and may be others ) have, we should use this fix until an API will be available.

shipit

henrygrover commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,229 @@
+        diff={}
+    )
+
+    # get the origional value for the given config option
```
Typo

itewk commented:
@yaacov @cben @elad661 I have updated my pull request. I have been able to get rid of the awful wait loop. With some digging I found `MiqServer.my_server.save!` which blocks. This solves the async issue. I also fixed the spelling issue @henrygrover found.

yaacov commented:
shipit 👍 

p.s.
we will replace this when we have an API , but until we do, this is fine by me.

itewk commented:
Alright. Fixed the new syntax issues. And also updated the `version_added` flag from 2.5 to 2.6. I guess I missed the 2.5 cut off, much sadness.

yaacov commented:
shipit

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,211 @@
+options:
+    name:
+        description:
+            - Name of the ManageIQ config element to modify.
```
Since typical use sets something several levels deep anyway, consider not treating top level specially.
E.g. instead of:
```
name: foo
value:
  bar:
    baz: 42
```
it could be more uniform to have user write:
```
settings:  # or `values:` or whatever. need something to separate from non-setting params like vmdb_path.
  foo:
    bar:
      baz: 42
```
This would also allow touching several top-level trees at once.

The drawback I see is this would inflate the return value & diff from a single subtree (which can already be big) to whole settings tree, which would be cumbersomely large :-(
So I'm not sure it's worth it, just an idea, your call...

itewk replied:
I think I like keeping it broken up. But i think it more comes down to personal preference. I could go either way if you all wanted to see it switched to the entire tree I will update.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,211 @@
+            - Name of the ManageIQ config element to modify.
+        required: True
+    value:
+        description: Dictionary to set as the value for the ManageIQ config option. Any values not given will be left at current value.
```
It would be nice to somehow clarify it's a deep merge.  It's not obvious "Any not given will be left at current value" extends down, one could read this as you may omit 2nd level keys but have to provide whole subtrees from 3rd level and below.

itewk replied:
Makes sense.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,211 @@
+    returncode, out, err = module.run_command([
+        "rails",
+        "r",
+        "puts MiqServer.my_server.get_config(:%s).config.to_json" % (name)
```
FYI, in case you switch to working on whole config, `get_config.config` gives you the whole tree.  (`get_config(:foo).config` is generally equivalent to `get_config.config[:foo]`.)
Also, VMDB::Config is deprecated.  May be better to use `::Settings.to_hash`.

itewk replied:
@cben I am not familiar with `::Settings.to_hash` could you point me where I would need to look for an example or replacing current `MiqServer.my_server.get_config` call with your suggestion?

carbonin replied:
+1 for @cben's suggestion here.

I would think `::Settings.to_hash` or `MiqServer.my_server.settings_for_resource` would be preferred here. I'm not sure which is the better way to go .. @Fryguy ?

Fryguy replied:
I agree with @cben here.  VMDB::Config is deprecated, and really only was kept around for some UI code that was using it.  If you are basically on the server, it's better to use `::Settings.to_hash` (or more specifically `::Settings.to_hash.to_json`)

elad661 commented:
shipit

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,211 @@
+        (update_value_rc, update_value_out, update_value_err) = module.run_command([
+            "rails",
+            "r",
+            "MiqServer.my_server.set_config(:%s => JSON.parse('%s')); MiqServer.my_server.save!" % (module.params['name'], json.dumps(expected_value))
```
I'm worried about Ruby quoting.  If any setting contains a single quote, this will break.
What I suggest is passing via ARGV.  `rails r` leaves in ARGV any args following the ruby fragment, e.g. this works:
```
# The \' escaping below is only to type this via shell;
# run_command() takes an array of strings and I think this way could pass arbitrary JSON with no other escaping at all.
rails r 'puts MiqServer.my_server.set_config(JSON.parse(ARGV.last))' '{"goodbye": "Qapla\', Worf"}'
```

itewk replied:
nifty trick. I like it.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,211 @@
+        (update_value_rc, update_value_out, update_value_err) = module.run_command([
+            "rails",
+            "r",
+            "MiqServer.my_server.set_config(:%s => JSON.parse('%s')); MiqServer.my_server.save!" % (module.params['name'], json.dumps(expected_value))
```
I'm surprised `set_config` works with just one subtree. Experimentally, this works, but I'm not sure it's safe — why doesn't it blow away all the other subtrees?
This boils down to `Vmdb::Settings.save!(MiqServer.my_server, name => new_value)`.
https://github.com/ManageIQ/manageiq/blob/d4a91f54341/lib/vmdb/settings.rb#L48-L56
my head hurts thinking how all this diffing/merging adds up :-(  I suspect it's simply that SettingsChange records can't represent deletions?
Anyway, I'd feel better if you passed whole tree (merged with desired changes) instead of just one.
BTW, you can `.deep_merge(...)` on ruby side too, if you prefer.

P.S. I'm not sure `my_server.save!` is necessary (though harmless).  `set_config` calls `Vmdb::Settings.save!` which does the work of writing SettingsChange records; `my_server.save!` is just ActiveRecord to save that MiqServer record, if modified — and testing in rails console shows `set_config` did not modify it.

itewk replied:
@cben the reason i added `MiqServer.my_server.save!` is that without it the configuration change is asynchronous. That is calling `MiqServer.my_server.set_config` puts the operation in a queue. Previous iterations of this module did a wait loop checking for the change to actually take affect. It was a mess. Adding in the `MiQServer.my_server.save!` from my testing is blocking and thus eliminated the need for the crazy wait loop

itewk replied:
With some testing using the Vmdb::Settings.save! looks to be blocking, so that is sweet.

cben replied:
Sorry, I think I mislead you, you may need to go back to `set_config` (or do what it does).
I couldn't understand where you sync/async difference comes from.
`set_config`, besides calling `Vmdb::Settings.save!`, ends up running `MiqServer.my_server.reload_all_server_settings`.  That was a no-op in my testing because the server was down, I only run `rails c`, and I didn't understand how it can help because `rails c` runs in its own process.
But it *can* help — when you have a running server, it *enqueues* a "reload_settings" call. 
Unfortunately, that's async, I don't see how anything can affect other processes synchronously, except it may succeed by chance... :(

I didn't look at your previous loop, but if that's the only direction, I propose dropping it. I'm not sure module should block until all workers saw the reload.  If modules wrote to DB, and queued reload, IMHO that's good enough.

cben reviewed (changes requested):
Sorry for silence.  Yeah, wishing for an API didn't make 
 it materialize :-), let's have a hackish module for now :+1:
This is getting better.
Lots of comments, but the only one I consider blocker is about Ruby quoting — I believe current code will blow up if the config subtree you modify contains single quote `'` chars anywhere.

cben commented:
@elad661 @yaacov if you prefer to merge and iterate, I can add shipit too, let me know.  (not sure if ansibot will now merge with your 2 shipits + mine "request changes", but if it does that's OK)

elad661 commented:
@cben I have no strong preference. Depends on where the cutoff time for 2.6 is. If there's enough time to make sure it gets in good shape before the cutoff, then it's fine to merge and iterate later.
I think @itewk should decide.

yaacov commented:
I'm in favour of merging and then iterating, and fixing this issues and other raised when people start using this :-)

@itewk it's up to you to decide what you want to fix from @cben 's list , once you give the OK @cben will re-review.

cben commented:
shipit

yaacov commented:
@cben I think you also need to remove the chenge-request ?

itewk committed:
manageiq_config.py - replace depreciated MiqServer.my_server.set_config with Vmdb::Settings.save! and protect against single quotes in value by using ARGV

itewk committed:
manageiq_config.py - update documenation

itewk commented:
@cben thank you for the review :)

I got through a couple of your recommendations. Wasn't sure if you wanted to include them in this PR at this point or if I should put in a second PR.

https://github.com/itewk/ansible/commit/9fcc25e7be31f60be44cc93a6e0b781edc7fc49c
https://github.com/itewk/ansible/commit/ff29c475b9728586ab0aa0a614c05ba87c6fb8d6

Let me know.

cben commented:
This PR vs followup — whatever you prefer.

cben commented:
Hmm, I suppose the reason ansibot didn't merge already is `new_module`, this probably needs a review from a core Ansible maintainer?  If so, you're better off continuing in this PR.

itewk commented:
@cben and @yaacov merged in the recommendations. Thanks again for all the reviews, comments, suggestions, etc.

yaacov commented:
shipit 
:-)

itewk commented:
@cben how do the changes look to you?

cben commented:
See my comment https://github.com/ansible/ansible/pull/35182#discussion_r167436078 about reloading config, I'm afraid I mislead you.
Anyway LGTM.
shipit

itewk commented:
@cben i went through a whole new test cycle after updating to use `Vmdb::Settings.save!(MiqServer.my_server, ...)` that call did not have the async issue I had with `MiqServer.my_server.set_config` that required adding in the `MiqServer.my_server.save!` to do the blocking. So unless you are speaking of something else, I don't think we have any issues.

itewk commented:
@cben and @yaacov now that we have the nifty `shipit` tag does the ansibot take over from there?

elad661 commented:
@itewk unfortunately not. Since this is a new module pull request, it needs to be reviewed by a core ansible maintainer (in addition to module maintainers like us) before it can be merged.

yaacov commented:
> @cben and @yaacov now that we have the nifty shipit tag does the ansibot take over from there?

@resmo ?

resmo commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,213 @@
+        argument_spec=dict(
+            name=dict(type='str', required=True),
+            value=dict(type='dict', required=False, default={}),
+            vmdb_path=dict(type='str', required=False, default='/var/www/miq/vmdb'),
```
Please verify if type='path' can be used here, also see https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1938

resmo commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,213 @@
+    #       to update other defaults on the first update of the configuration after
+    #       and upgrade. Therefor, recreating the expected value from the updated current
+    #       value protects against that.
+    current_value = get_manageiq_config_value(module, module.params['name'])
```
I noticed, there is a raise exception in this function, but not try/except here, (there is one on Line 151). also add try catch here?

Fryguy commented:
Sorry all...I didn't see the ping until now.  We *do* have an API for settings, so I'm a little confused.  https://github.com/ManageIQ/manageiq-api/pull/275 was merged on Jan 15, and it was released as part of Gaprindashvili-1.

itewk commented:
@Fryguy well feb 15th is relatively new. i guess ill have to re-write against the API and keep this version for everyone on older versions of manageiq/cfme who wont have access to the API

Fryguy commented:
We tried to backport to Fine (a.k.a. CF 5.8), but it was built on a lot of refactoring of the API code, and thus was too hard to backport.

Fryguy commented:
@bdunne or @carbonin Do you think you can take a look here and see if everything looks reasonable?

carbonin commented on lib/ansible/modules/remote_management/manageiq/manageiq_config.py:
```diff
@@ -0,0 +1,213 @@
+        (update_value_rc, update_value_out, update_value_err) = module.run_command([
+            "rails",
+            "r",
+            "Vmdb::Settings.save!(MiqServer.my_server, :%s => JSON.parse(ARGV.last))" % (module.params['name']),
```
I think it would be better to use `MiqServer.my_server.add_settings_for_resource(…)` here. This will queue a settings reload for all workers running on the target server after save.

cben commented:
BTW, settings REST API currently can't round-trip symbol values without turning them into strings, which breaks all kinds of code that expected symbols
https://bugzilla.redhat.com/show_bug.cgi?id=1558031

Does this module have similar problem?  It could in principle bypass it, as it goes through Ruby not REST API, but I think this implementation uses JSON between Ruby and Python and would also loss symbol/string distinction.

itewk commented:
@cben in the couple months I have been using this with clients I have not run into any issues with the symbols. As you say, since going direct to ruby and running everything through parsers I think the issue going on with the REST API is avoided. If you have a specific scenario that you know doesn't work with the REST API could try it with this module.

cben commented:
Hi, didn't realize this is still stuck unmerged.

The symbols issue is mostly (but [not completely](https://github.com/ManageIQ/manageiq/issues/17201#issuecomment-385925194)) solved on master, but I don't think that's going to be backported, and the whole point of this module not using API was working with older manageiq.

To confirm symbols are not being converted to strings, could you run:
```
bundle exec rails runner 'puts Settings.to_hash.to_yaml'
```
before and after using the module, and diff the outputs?
